sqlglot/typing/snowflake.py

```diff
@@ -81,6 +82,230 @@ def _annotate_decode_case(self: TypeAnnotator, expression: exp.DecodeCase) -> ex
+        return exp.DataType(
+            this=exp.DataType.Type.DECIMAL,
+            expressions=[
+                exp.DataTypeParam(this=exp.Literal.number(precision)),
+                exp.DataTypeParam(this=exp.Literal.number(scale)),
+            ],
+        )
+
+    if expression.type.is_type(*exp.DataType.INTEGER_TYPES) and not expression.type.expressions:
+        return exp.DataType(
+            this=exp.DataType.Type.DECIMAL,
+            expressions=[
+                exp.DataTypeParam(this=exp.Literal.number(38)),
+                exp.DataTypeParam(this=exp.Literal.number(0)),
+            ],
+        )
+
+    return expression.type
+
+
+def _coerce_two_parameterized_types(
+    type1: exp.DataType, p1: int, s1: int, type2: exp.DataType, p2: int, s2: int
+) -> t.Optional[exp.DataType]:
+    """
+    Coerce two parameterized numeric types using Snowflake's type coercion rules.
+
+    Rules:
+    - If p1 >= p2 AND s1 >= s2: return type1
+    - If p2 >= p1 AND s2 >= s1: return type2
+    - Otherwise: return NUMBER(min(38, max(p1, p2) + |s2 - s1|), max(s1, s2))
+    """
+    if p1 >= p2 and s1 >= s2:
+        return type1.copy()
+
+    if p2 >= p1 and s2 >= s1:
+        return type2.copy()
+
+    result_scale = max(s1, s2)
+    result_precision = min(38, max(p1, p2) + abs(s2 - s1))
+
+    return exp.DataType(
+        this=type1.this,
+        expressions=[
+            exp.DataTypeParam(this=exp.Literal.number(result_precision)),
+            exp.DataTypeParam(this=exp.Literal.number(result_scale)),
+        ],
+    )
+
+
+def _coerce_parameterized_numeric_types(
+    types: t.List[t.Optional[exp.DataType]],
+) -> t.Optional[exp.DataType]:
+    """
+    Generalized function to coerce multiple parameterized numeric types.
+    Applies Snowflake's coercion logic pairwise across all types.
+    """
+    if not types:
+        return None
+
+    result_type = None
+
+    for current_type in types:
+        if not current_type:
+            continue
+
+        if result_type is None:
+            result_type = current_type
+            continue
+
+        if not _is_parameterized_numeric(result_type) or not _is_parameterized_numeric(
+            current_type
+        ):
+            return None
+
+        p1, s1 = _extract_type_precision_scale(result_type)
+        p2, s2 = _extract_type_precision_scale(current_type)
+
+        if p1 is None or s1 is None or p2 is None or s2 is None:
+            return None
+
+        result_type = _coerce_two_parameterized_types(result_type, p1, s1, current_type, p2, s2)
+
+    return result_type
+
+
+T = t.TypeVar("T", bound=exp.Expression)
+
+
+def _apply_numeric_coercion(
+    self: TypeAnnotator,
+    expression: T,
+    expressions_to_coerce: t.List[exp.Expression],
+) -> t.Optional[T]:
+    if any(_is_float(e.type) for e in expressions_to_coerce):
+        self._set_type(expression, exp.DataType.Type.FLOAT)
+        return expression
```

Comment:
This feels like something that should be handled by a `COERCES_TO` override for Snowflake, i.e. `NUMBER` always coerces to `FLOAT`:

```
(sqlglot) ➜  sqlglot git:(main) snow sql -q "select system\$typeof(coalesce(cast(1 as number(4, 2)), cast(1 as float)))"
select system$typeof(coalesce(cast(1 as number(4, 2)), cast(1 as float)))
+--------------------------------------------------------------------+
| SYSTEM$TYPEOF(COALESCE(CAST(1 AS NUMBER(4, 2)), CAST(1 AS FLOAT))) |
|--------------------------------------------------------------------|
| FLOAT[DOUBLE]                                                      |
+--------------------------------------------------------------------+
```